python/truelayer_signing/utils.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+# std imports
+import base64
+import json
+from enum import Enum
+from typing import Dict, Mapping, MutableMapping, Optional, Tuple, TypeVar
+
+T = TypeVar('T', bound='TlJwsBase')
+
+
+class HttpMethod(str, Enum):
+    POST = "POST"
+    GET = "GET"
+    PATCH = "PATCH"
+    PUT = "PUT"
+    DELETE = "DELETE"
+
+
+class TlJwsBase:
+    pkey: str
+    http_method: HttpMethod
+    path: str
+    headers: MutableMapping[str, str]
+    body: str
+
+    def __init__(
+        self,
+        pkey: str,
+        http_method: HttpMethod = HttpMethod.POST,
+        path: str = "",
+        headers: Optional[Dict[str, str]] = None,
+        body: str = ""
+    ):
+        self.pkey = pkey
+        self.http_method = http_method
+        self.path = path
+        self.headers = {} if headers is None else headers
+        self.body = body
+
+    def set_method(self: T, http_method: HttpMethod) -> T:
+        """
+        Add the request method, defaults to `"POST"` if unspecified.
+        """
+        self.http_method = http_method
+        return self
+
+    def set_path(self: T, path: str) -> T:
+        """
+        Add the request absolute path starting with a leading `/` and without
+        any trailing slashes.
+        """
+        self.path = path
+        return self
+
+    def add_headers(self: T, headers: MutableMapping[str, str]) -> T:
```

Comment:
Since ordering matters, what do you think about making headers an `Iterable[Tuple[str, str]]` or `Iterator[...]`? Users would have to call `.items()` when passing `dict`s but then at least they're making a concious choice of relying on the unspecified ordering of `dict` keys